File: leetcode/two-pointers/88.Merge-Sorted-Array/88.Merge-Sorted-Array.go
package leetcode

import "sort"

// 方法1：直接合并后排序
func merge1(nums1 []int, m int, nums2 []int, n int)  {
    copy(nums1[m:], nums2)
	sort.Ints(nums1)
}
/*
最直观的方法是先将数组nums2放进数组nums1的尾部，
然后直接对整个数组进行排序。

- 时间复杂度：O((m+n)log(m+n))。
排序序列长度为 m+n，套用快速排序的时间复杂度即可，平均情况为 O((m+n)log(m+n))。

- 空间复杂度：O(log(m+n))。
排序序列长度为 m+n，套用快速排序的空间复杂度即可，平均情况为 O(log(m+n))。
*/

// 方法二：双指针
func merge2(nums1 []int, m int, nums2 []int, n int)  {
    sorted := make([]int, 0, m + n)
	p1, p2 := 0, 0
	for {
		if p1 == m {
			sorted = append(sorted, nums2[p2:]...)
			break
		}
		if p2 == n {
			sorted = append(sorted, nums1[p1:]...)
			break
		}
		if nums1[p1] < nums2[p2] {
			sorted = append(sorted, nums1[p1])
			p1++
		} else {
			sorted = append(sorted, nums2[p2])
			p2++
		}
	}
	copy(nums1, sorted)
}

/*
将两个数组看作队列，
每次从两个数组头部取出比较小的数字放到结果中

- 时间复杂度：O(m+n)。
指针移动单调递增，最多移动 m+n 次，因此时间复杂度为 O(m+n)。

- 空间复杂度：O(m+n)。
需要建立长度为 m+n 的中间数组 sorted。
*/

// 方法三 逆向双指针
func merge(nums1 []int, m int, nums2 []int, n int)  {
    for p := m + n; m > 0 && n > 0; p-- {
		if nums1[m-1] <= nums2[n-1] {
			nums1[p-1] = nums2[n-1]
			n--
		} else {
			nums1[p-1] = nums1[m-1]
			m--
		}
	}
	for ; n > 0; n-- {
		nums1[n-1] = nums2[n-1]
	}
}
/*
为了不大量移动元素，就要从2个数组长度之和的最后一个位置开始，
依次选取两个数组中大的数，从第一个数组的尾巴开始往头放，
只要循环一次以后，就生成了合并以后的数组了。

时间复杂度：O(m+n)。
指针移动单调递减，最多移动 m+n 次，因此时间复杂度为 O(m+n)。

空间复杂度：O(1)。
直接对数组nums1原地修改，不需要额外空间。
*/

File: leetcode-cpp/88.Merge-Sorted-Array/88.Merge-Sorted-Array.cpp
/**
 * @file 88.Merge-Sorted-Array.cpp
 * @author your name (you@domain.com)
 * @brief leetcode-88.Merge-Sorted-Array
 * @version 0.1
 * @date 2023-08-13
 * 
 * @copyright Copyright (c) 2023
 * 给你两个按 非递减顺序 排列的整数数组 nums1
 *  和 nums2，另有两个整数 m 和 n ，
 * 分别表示 nums1 和 nums2 中的元素数目。

请你 合并 nums2 到 nums1 中，使合并后的数组
同样按 非递减顺序 排列。

注意：最终，合并后数组不应由函数返回，
而是存储在数组 nums1 中。为了应对这种情况，
nums1 的初始长度为 m + n，其中前 m 个元素
表示应合并的元素，后 n 个元素为 0 ，应忽略。
nums2 的长度为 n 。

示例 1：

输入：nums1 = [1,2,3,0,0,0], m = 3, 
nums2 = [2,5,6], n = 3
输出：[1,2,2,3,5,6]
解释：需要合并 [1,2,3] 和 [2,5,6] 。
合并结果是 [1,2,2,3,5,6] ，其中斜体加粗标注
的为 nums1 中的元素。

示例 2：

输入：nums1 = [1], m = 1, 
nums2 = [], n = 0
输出：[1]
解释：需要合并 [1] 和 [] 。
合并结果是 [1] 。

示例 3：

输入：nums1 = [0], m = 0, 
nums2 = [1], n = 1
输出：[1]
解释：需要合并的数组是 [] 和 [1] 。
合并结果是 [1] 。
注意，因为 m = 0 ，所以 nums1 中没有元素。
nums1 中仅存的 0 仅仅是为了确保合并结果
可以顺利存放到 nums1 中。

提示：

nums1.length == m + n
nums2.length == n
0 <= m, n <= 200
1 <= m + n <= 200
-10^9 <= nums1[i], nums2[j] <= 10^9
 

进阶：你可以设计实现一个时间复杂度为 O(m + n) 的算法解决此问题吗？
 */

#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    // 大佬解法：从后面开始确定
    void merge(vector<int>& nums1, int m, vector<int>& nums2, int n) {
        int i = nums1.size() - 1;
        m--;
        n--;
        while (n >= 0) {
            while (m >= 0 && nums1[m] > nums2[n]) {
                swap(nums1[i--], nums1[m--]);
            }
            swap(nums1[i--], nums2[n--]);
        }
    }

    // 太累了，想休息了，08-14再完成leetcode官方三种方法吧
};

int main() {
    vector<int> nums1 = {1,2,3,0,0,0};
    vector<int> nums2 = {2,5,6};

    Solution().merge(nums1, 3, nums2, 3);

    // 打印nums1
    for (auto& num : nums1) {
        cout << num << endl;
    }

    return 0;
}